### ISSUE TITLE
Virtual bytes (sat/vB) should be used for fee calculation

### ISSUE THREAD

**Issue by softsimon** [BlueWallet/BlueWallet#2785]
### Current behavior
When making a transaction and selecting fee, BlueWallet uses satoshis/byte, so when I made a test transaction today and picked 19 sat/byte, it resulted in 33.4 sat/vB and thus more fee than I expected.

Even when clicking the built in hard coded "View in Block explorer" you are taken to Blockstream dot info that display **33.4 sat/vB** and doesn't match what I just selected in fee level. I made the deposit to Electrum Wallet which also display **33.4 sat/vB**. This causes confusion.

### Expected behavior
Satoshis per vbyte is the new standard to be used since Segwit was introduced. This unifies every measurement into a single backwards compatible one that works both for legacy and segwit-transactions.

Here @sipa confirms that the Bitcoin Core reference implementation uses vbytes as default:

> [Bitcoin Core] is of course already using vbytes; anything else doesn't make sense. Almost everywhere Bitcoin Core reports transaction sizes, they're reported in vbytes

Source: https://bitcoin.stackexchange.com/questions/87291/is-theres-anyway-to-get-fee-estimation-with-vbyte-instead-of-kb-using-estimate

I have seen multiple modern wallets like Electrum, Sparrow and Green using only vBytes, and I am developing the https://github.com/mempool/mempool project which **solely** uses vBytes for its fee estimation and presentation. 

BlueWallet should follow suit.

### Tested on
BlueWallet version: 6.06
Device: Google Pixel 4

**Comment by limpbrains** [comment]
I agree that it is confusing and BW should use sat/vbyte everywhere. And show it explicitly.

**Comment by ncoelho** [comment]
Was looking into this yesterday actually. And I do agree that we should move towards vb, specially to make it less confusing across clients.

Fun fact:
Doing the same transaction on electrum and bluewallet is specially confusing and users are bringing this up. Although electrum is using the vsize on the background it is displaying bytes on the frontend. 

Same transaction:
Blue | Electrum
------------ | -------------
<img width="371" alt="Screenshot 2021-03-12 at 16 28 47" src="https://user-images.githubusercontent.com/1238429/110961713-62043f80-8350-11eb-9035-5b7186bd65d9.png"> | <img width="1193" alt="Screenshot 2021-03-12 at 16 28 37" src="https://user-images.githubusercontent.com/1238429/110961717-64669980-8350-11eb-9c62-b26b6a735ff5.png">

At the same time, here we are, introducing one more concept on the user facing layer, that probably 10 people in the world now how to explain properly  🤦

**Comment by bholm** [comment]
Any update to when this will be implemented? I intended to make a transaction for 1 sat/vB tonight with the mempool being empty but it ended up being 2.1 sat/vB. Normally I would work around this issue by making the fee smaller but there is no way to do less than 1 sat/B.

**Comment by jaonoctus** [comment]
We need to [Get recommended fees](#get-recommended-fees) in sat/vbyte instead of sat/byte, then [Create transaction](#create-transaction) with the new fee rate to the [coinselection](https://github.com/bitcoinjs/coinselect).

The thing is that the current coinselection algo doest not support segwit/vbyte.

- This is a possible fix for coinselection algo: https://github.com/bitcoinjs/coinselect/pull/63

## Get recommended fees

### Code

https://github.com/BlueWallet/BlueWallet/blob/2fb16c7bad9d1b8c16a53f4c5db8c60c0e2d2380/blue_modules/BlueElectrum.js#L766-L786

### Flow

![getrecommendedfees](https://user-images.githubusercontent.com/7918067/132135164-ce9fd41d-befd-4b03-9382-b9b95eb028fc.jpeg)

```mermaid
sequenceDiagram
title: GET RECOMMENDED FEES
details.js->>NetworkTransactionFees: recommendedFees()
NetworkTransactionFees->>BlueElectrum: estimateFees()
BlueElectrum->>ElectrumClient: mempool_getFeeHistogram()
ElectrumClient-->>BlueElectrum: histogram
BlueElectrum->>ElectrumClient: blockchainEstimatefee()
Note over ElectrumClient: bitcoin core
ElectrumClient->>BlueElectrum: estimateFee for 1, 18 and 144 blocks
BlueElectrum->>BlueElectrum: calcEstimateFeeFromFeeHistorgam()
Note over BlueElectrum: mempool
BlueElectrum-->>NetworkTransactionFees: {fast, medium, slow}
NetworkTransactionFees-->>details.js: {fast, medium, slow}
```

## Create transaction

### Code

https://github.com/BlueWallet/BlueWallet/blob/2fb16c7bad9d1b8c16a53f4c5db8c60c0e2d2380/class/wallets/abstract-hd-electrum-wallet.js#L856

https://github.com/BlueWallet/BlueWallet/blob/2fb16c7bad9d1b8c16a53f4c5db8c60c0e2d2380/class/wallets/legacy-wallet.ts#L377-L406

### Flow

> incomplete

![createtransaction](https://user-images.githubusercontent.com/7918067/132135330-35f0e2e8-36fe-4484-bd1b-5ab745f19adf.jpeg)

```mermaid
sequenceDiagram
title: CREATE TRANSACTION
details.js->>details.js: createPsbtTransaction()
details.js->>wallet: createTransaction()
wallet->>coinselect: select coins using feeRate
coinselect-->>wallet: {inputs, outputs, fee}
wallet->>wallet: new bitcoin.Psbt()
wallet-->>details.js: { tx, inputs, outputs, fee, psbt }
```

**Comment by jaonoctus** [comment]
Possible fix here: jaonoctus/coinselect#1

**Comment by jaonoctus** [comment]
# Change

> With this change, I got ~40% lower fees.

- Why 27 and 22: https://github.com/bitcoinjs/coinselect/pull/63#issuecomment-917310462

```ts
const uxtosSegwitCompatible = utxos.map(utxo => {
  const u = (utxo as unknown) as Utxo;
  if (u.address.startsWith('bc1')) {
    // TODO: check if its P2WPKH or P2WSH
    return {
      ...utxo,
      script: {
        length: 27,
      },
    };
  }

  return utxo;
});

const targetsSegwitCompatible = targets.map(target => {
  if (target.address.startsWith('bc1')) {
    // TODO: check if its P2WPKH or P2WSH

    return {
      ...target,
      script: {
        length: 22,
      },
    };
  }

  return target;
});

const { inputs, outputs, fee } = algo(uxtosSegwitCompatible, targetsSegwitCompatible, feeRate);
```

# Tests

* feeRate: 1

## 1 input, 2 outputs

> -37%

type | fee | inputs | outputs
-- | -- | -- | --
sat/vB | 143 | 291186 | 100000<br>191043 (change)
sat/B | 226 | 291186 | 100000<br>190960 (change)

## 1 input, 3 outputs

> -33%

type | fee | inputs | outputs
-- | -- | -- | --
sat/vB | 174 | 291186 | 50000<br>50000<br>191012 (change)
sat/B | 260 | 291186 | 50000<br>50000<br>190926 (change)

## 2 inputs, 2 outputs

> -44%

type | fee | inputs | outputs
-- | -- | -- | --
sat/vB | 211 | 291186<br>175262 | 400000<br>66237 (change)
sat/B | 374 | 291186<br>175262 | 400000<br>66074 (change)

## 2 inputs, 3 outputs

> -41%

type | fee | inputs | outputs
-- | -- | -- | --
sat/vB | 242 | 291186<br>175262 | 200000<br>200000<br>66206 (change)
sat/B | 408 | 291186<br>175262 | 200000<br>200000<br>66040 (change)

## 4 inputs, 3 outputs

> -46%

type | fee | inputs | outputs
-- | -- | -- | --
sat/vB | 378 | 291186<br>175262<br>100000<br>56000 | 400000<br>200000<br>22070 (change)
sat/B | 704 | 291186<br>175262<br>100000<br>56000 | 400000<br>200000<br>21744 (change)
